### PR TITLE
initial release of nat64 module by Vitold

### PR DESCRIPTION
**Pull request by dshamaev-intermedia**
- introduced nat64 module that is targeted to solve network traversal issues with connecting to HPBX server that has only public IPv4 address and pjsip module is placed in IPv6-only network
- bumped versions

### CODE REVIEW

File: pjnath/build/Makefile [hunk 1]
```diff
@@ -162,6 +162,4 @@ depend:
 	$(MAKE) -f $(RULES_MAK) APP=PJTURN_SRV app=pjturn-srv $@
 	echo '$(BINDIR)/$(PJNATH_TEST_EXE): $(LIBDIR)/$(PJNATH_LIB) $(PJLIB_UTIL_LIB) $(PJLIB_LIB)' >> .pjnath-test-$(TARGET_NAME).depend
 	echo '$(BINDIR)/$(PJTURN_CLIENT_EXE): $(LIBDIR)/$(PJNATH_LIB) $(PJLIB_UTIL_LIB) $(PJLIB_LIB)' >> .pjturn-client-$(TARGET_NAME).depend
-	echo '$(BINDIR)/$(PJTURN_SRV_EXE): $(LIBDIR)/$(PJNATH_LIB) $(PJLIB_UTIL_LIB) $(PJLIB_LIB)' >> .pjturn-srv-$(TARGET_NAME).depend
-
```
Comment (anikitin-intermedia) [inline pjnath/build/Makefile:166]: Unnecessary changes

**Review by anikitin-intermedia — APPROVED** [review]

File: pjnat64/src/pjnat64/pj-nat64.c [hunk 2]
```diff
@@ -0,0 +1,1248 @@
...
+#include <pj/types.h>
+
+#include <pjsip-ua/sip_inv.h>
+#include <pjsip/sip_types.h>
+
+#include <pjsua.h>
+
+#include <pjnat64/pj-nat64.h>
+
+#define THIS_FILE "pj_nat64.c"
```
Comment (anikitin-intermedia) [inline pjnat64/src/pjnat64/pj-nat64.c:11]: `"pj-nat64.c"`

File: pjnat64/src/pjnat64/pj-nat64.c [hunk 3]
```diff
@@ -0,0 +1,1248 @@
...
+
+#define MODULE_VERSION "$Id: 497af2bf5811a519a87deb9821451887ec3200e6 $"
```
Comment (anikitin-intermedia) [inline pjnat64/src/pjnat64/pj-nat64.c:13]: I think this is not needed anymore, is not?

File: ios_framework/config_site.h [hunk 4]
```diff
@@ -1,5 +1,7 @@
 #define PJ_CONFIG_IPHONE 1
+#define PJ_CONFIG_IPHONE 1
```
Comment (magdalenapopovic) [inline ios_framework/config_site.h:2]: Doubled?

File: ios_framework/config_site.h [hunk 5]
```diff
@@ -1,5 +1,7 @@
...
 #define PJ_HAS_IPV6 1
+// #define PJ_LOG_MAX_LEVEL 6
 #define PJMEDIA_HAS_BCG729 1
```
Comment (magdalenapopovic) [inline ios_framework/config_site.h:5]: Please remove.

File: pjnat64/src/pjnat64/pj-nat64.c [hunk 6]
```diff
@@ -0,0 +1,1409 @@
...
+    if ((attr == NULL) || (attr->name.slen == 0)) {
+        return;
+    }
+
+    PJ_LOG(4, (THIS_FILE, "Process SDP attribute %.*s", attr->name.slen, attr->name.ptr));
+
+    if (strncmp(attr->name.ptr, rtcp_name.ptr, rtcp_name.slen) == 0) {
+        s2 = pjmedia_sdp_attr_get_rtcp(attr, &ra);
+        if (s2 != PJ_SUCCESS) {
+            PJ_LOG(1, (THIS_FILE, "Error: Patch SDP attribute 'rtcp'"));
```
Comment (magdalenapopovic) [inline pjnat64/src/pjnat64/pj-nat64.c:110]: Can be a log, not an error log. Rtcp port not usually present in SDP.

File: pjnat64/src/pjnat64/pj-nat64.c [hunk 7]
```diff
@@ -0,0 +1,1409 @@
...
+            ));
+            sip_uri->host.slen = map_addr.slen;
+            sip_uri->host.ptr = map_addr.ptr;
+            /* Update patch count */
+            patch_count += 1;
+        }
+    }
+#endif
+
+    /* Step 2. Patch R-URI on INVITE/ACK/BYE request with server IPv4 address */
```
Comment (magdalenapopovic) [inline pjnat64/src/pjnat64/pj-nat64.c:936]: Do you explicitly check INVITE/ACK/BYE? You can do this for every sip request.

**Review by magdalenapopovic — APPROVED** [review]
Great work Vitold :)

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 6 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c:110]: This is a trick to parse into `ra` struct `rtcp` attribute. Checking that we have an `rtcp` provide two lines earlier at `107`. In general, I think that there is probably a standard `rtcp` parsing function, but the documentation is the first thing I found.

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 7 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c:936]: Yep. I think we can remove checks method here.

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 3 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: @dshamaev-intermedia remove?

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 2 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: We can replace it with `__FILE__` this will generally give an strict file name. Let's do that?

Comment (anikitin-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: No, we can't, because __FILE__ returns full path as I understand

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 3 above)
Comment (dshamaev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: yes, after we will merg it here, i think it's not relevant anymore

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 2 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: Ok. Renamed with use minus.

File: pjnat64/src/pjnat64/pj-nat64.c (hunk 3 above)
Comment (vsedyshev-intermedia) [reply pjnat64/src/pjnat64/pj-nat64.c]: Removed.

**Review by vsedyshev-intermedia — APPROVED** [review]
Look like demo ready.
